cedar-policy-validator/src/schema_file_format.rs

```diff
@@ -441,11 +462,100 @@ impl<N: std::fmt::Display> std::fmt::Display for ActionEntityUID<N> {
 }
 
 impl ActionEntityUID<RawName> {
-    /// Prefix unqualified entity and common type references with the namespace they are in
-    pub fn qualify_type_references(self, ns: Option<&Name>) -> ActionEntityUID<Name> {
+    /// (Conditionally) prefix this action entity UID's typename with the given namespace
+    pub fn conditionally_qualify_type_references(
+        self,
+        ns: Option<&Name>,
+    ) -> ActionEntityUID<ConditionalName> {
         ActionEntityUID {
             id: self.id,
-            ty: self.ty.map(|rname| rname.qualify_with(ns)),
+            ty: {
+                // PANIC SAFETY: this is a valid raw name
+                #[allow(clippy::expect_used)]
+                let raw_name = self
+                    .ty
+                    .unwrap_or(RawName::from_str("Action").expect("valid raw name"));
+                Some(raw_name.conditionally_qualify_with(ns, ReferenceType::Entity))
+            },
+        }
+    }
+
+    /// Unconditionally prefix this action entity UID's typename with the given namespace
+    pub fn qualify_with(self, ns: Option<&Name>) -> ActionEntityUID<Name> {
+        ActionEntityUID {
+            id: self.id,
+            ty: {
+                // PANIC SAFETY: this is a valid raw name
+                #[allow(clippy::expect_used)]
+                let raw_name = self
+                    .ty
+                    .unwrap_or(RawName::from_str("Action").expect("valid raw name"));
+                Some(raw_name.qualify_with(ns))
+            },
+        }
+    }
+}
+
+impl ActionEntityUID<ConditionalName> {
+    /// Convert this [`ActionEntityUID<ConditionalName>`] into an
+    /// [`ActionEntityUID<Name>`] by fully-qualifying its typename.
+    ///
+    /// `all_action_defs` needs to be the full set of all fully-qualified action
+    /// EUIDs that are defined in the schema (in all schema fragments).
+    /// This `ActionEntityUID<ConditionalName>` must resolve to one of the
+    /// `all_action_defs` or else it throws `UndeclaredActionsError`.
+    pub fn fully_qualify_type_references(
+        self,
+        all_action_defs: &HashSet<EntityUID>,
+    ) -> std::result::Result<EntityUID, ActionResolutionError> {
+        for possibility in self.possibilities() {
+            let possibility = possibility.into();
+            if all_action_defs.contains(&possibility) {
+                return Ok(possibility);
+            }
+        }
+        return Err(ActionResolutionError(nonempty!(self)));
+    }
+
+    /// Get the possible fully-qualified [`ActionEntityUID<Name>`]s which this
+    /// [`ActionEntityUID<ConditionalName>`] might resolve to, in priority order
+    /// (highest-priority first).
+    pub(crate) fn possibilities(&self) -> impl Iterator<Item = ActionEntityUID<Name>> + '_ {
+        // PANIC SAFETY: by INVARIANT on self.ty
+        #[allow(clippy::expect_used)]
+        let ty = self.ty.as_ref().expect("by INVARIANT on self.ty");
```

Comment:
Perhaps encapsulate this with

```rust
impl ActionEntityUID<Name> {
  fn ty(&self) -> &Name {
    ...
  }
}
```

Reply:
Done, and improved all the invariant comments